Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2024 Richard Copley. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Richard Copley, Eric Wieser
+-/
+import Mathlib.LinearAlgebra.TensorAlgebra.Basis
+import Mathlib.Algebra.MonoidAlgebra.NoZeroDivisors
+
+/-!
+# Instances for the zero-product property in the tensor algebra of a free module
+
+This file provides the following instances:
+  * instNoZeroDivisorsMonoidAlgebraFreeMonoid
+      If `R` is a semiring with no zero-divisors and `κ` is any type
+      then `MonoidAlgebra (FreeMonoid κ) has no zero-divisors.
+  * FreeMonoid.instUniqueProds
+      The free monoid on any type has the UniqueProds property.
+      (This provides `NoZeroDivisors (MonoidAlgebra R (FreeMonoid κ))`, where
+      `R` is a semiring and `κ` is any type, via the instance
+      `MonoidAlgebra.instNoZeroDivisorsOfUniqueProds`.)
+  * FreeAlgebra.instNoZeroDivisors
+      If `R` is a commutative semiring with no zero-divisors and `X` is any type
+      then `MonoidAlgebra R X` has no zero-divisors.
+  * TensorAlgebra.instNoZeroDivisors
+      If `R` is a commutative semiring with no zero-divisors and `M` is a
+      free `Module` over `R` then `TensorAlgebra R M` has no zero-divisors.
+  * TensorAlgebra.instIsDomain
+      If `R` is an integral domain and `M` is a free module over `R` then
+      `TensorAlgebra R M` is a domain.
+-/
+
+-- Following https://math.stackexchange.com/a/75219/411545 by Mariano Suárez-Álvarez
+
+noncomputable section
+suppress_compilation
+
+instance FreeMonoid.instUniqueProds {κ : Type*} : UniqueProds (FreeMonoid κ) where
```

Comment:
Is the stronger `TwoUniqueProds` true?

Reply:
I just got back to my computer and it looks like I missed on a lot of fun here!

Anyway, I think that `TwoUniqueProds` should also hold: wherever the proof uses `max`, you could have used `min` just as well and this should be enough for `TwoUniqueProds` to go through.

Reply:
Choose two words of _minimal_ length ... let me check that.

Reply:
Yes, that'll work. Search-and-replace 'max' for 'min' and fix up a few things for an alternate proof of `UniqueProds`. That leaves a little housekeeping for edge cases. I'll have to leave that for another day, as it's getting late.

Reply:
> I'll have to leave that for another day
... or stay up late and do it badly ...
Needs golf.

Reply:
> I just got back to my computer and it looks like I missed on a lot of fun here!
Yeah!
> Anyway, I think that `TwoUniqueProds` should also hold: wherever the proof uses `max`, you could have used `min` just as well and this should be enough for `TwoUniqueProds` to go through.
Believe it or not, I didn't see this message until just now.